pkg/JWM-Format.md

```diff
@@ -0,0 +1,122 @@
+
+*enc*
+
+This field must contain the encryption algorithm of the scheme used to encrypt the ciphertext. It _MUST_ be a Authenticated encryption scheme which is defined within the JWM spec.
+
+Current proposed encryption schemes are:
+    1. ChaCha20Poly1305
+    2. XChaCha20Poly1305
+    3. Salsa20Poly1305
+    4. XSalsa20Poly1305
+
+*kid*
+
+This is a reference to the specific public key that was taken from the did-doc by the sender to encrypt the symmetrical key used to encrypt the ciphertext.
+
+*jwk*
+
+This field is used to specify which public key was used by the sender to encrypt the message. It _MUST_ be used when the DID_Doc of the receiver contains multiple messages to identify which public key was used by the sender to encrypt the message. It also _MUST_ be used when an ephemeral key is used to send the message, such as in SealedBox (AnonCrypt). The format of this _SHOULD_ be the public key base58 encoded.
+
+*encrypted_key*
+
+This is the key which has been encrypted using NaCl's crypto_box to encrypt the symmetrical key used to encrypt the ciphertext.
+
+# Encoding format
+
+Once the JWM has been properly filled out, the JSON message should be base64url encoded with each section being separated by a period. Examples of the encoding are provided below.
+
+# Example:
+[example]: #example
+
+### Compact Serialization Example 
+```json
+{
+    "header" : { "alg" : "NaCl_box", 
+                 "enc" : <AEAD algorithm name>,
+                 "kid" : <reference to the receiver public key as a DID>,
+                 "jwk" : <reference to the sender's public key>,
+                 "encrypted_key" : <output of a cryptobox function which is the secret key used to encrypt ciphertext>
+                },
+	"iv" : <Nonce>,
+    "ciphertext" : <message ciphertext>,
+    "tag" : <Authentication Tag from NaCl>
+}
+```
+
+### Encodes to:
+Base64URLEncode(header).
+Base64URLEncode(iv).
+Base64URLEncode(ciphertext).
+Base64URLEncode(tag)
+
+### JSON Serialization Example
+```json
+{
+    "recipient" : [
+        "header" : { "alg" : "NaCl_box", 
+                 "enc" : <AEAD algorithm name>,
+                 "kid" : <reference to the receiver public key as a DID>,
+                 "jwk" : <reference to the sender's public key>,
+                 "encrypted_key" : <output of a cryptobox function which is the secret key used to encrypt ciphertext>
+                },
+        "header" : { "alg" : "NaCl_box", 
+                 "enc" : <AEAD algorithm name>,
+                 "kid" : <reference to the receiver public key as a DID>,
+                 "jwk" : <reference to the sender's public key>,
+                 "encrypted_key" : <output of a cryptobox function which is the secret key used to encrypt ciphertext>
+                },
+    ],
+	"iv" : <Nonce>,
+    "ciphertext" : <message ciphertext>,
+    "tag" : <Authentication Tag from NaCl>
+}
+```
+
+### Encodes to:
+Base64URLEncode(header).
```

Comment:
So, which header (from the recipient block) goes here?

Reply:
I'm not sure what you mean. The idea of the recipient block is so that a sender can encrypt a message for multiple people. The receiver would find their kid and use the private key associated with the kid to decrypt `encrypted_key` which is essentially a crypto_box encryption of an ephemeral secret key. Once they decrypt and get the secret key and decrypt the ciphertext.

Reply:
There are two headers listed in the above example, but the compact form includes `Base64URLEncode(header)` but doesn't specify what to do if more than one header is present.

Did you mean `Base64URLEncode(recipient)` in the compact encoding

Reply:
yup, that's what I was thinking, I'll go and change it today.